### PR TITLE
RHAIFIRST-50: Add BranchResolver module for version-aware branch routing

### PR DESCRIPTION
**Pull request by EmilienM**
## Summary

- Add `agentic_ci.branch` module with `resolve_branch_from_jira()`, `validate_branch_exists()`, and `BranchResolutionError`
- Resolution chain: fixVersion → version_branches map → direct fixVersion → static branch field → None
- Validates resolved branch exists on remote via `git ls-remote`
- Imports `_validate_ref` from `git.py` (no duplication)

## Test plan

- [x] 28 unit tests covering all resolution paths, edge cases, and validation
- [x] `tox -e py313 -e lint -e check-format` all green

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * CI now derives target branches from Jira fix versions and optionally verifies the branch exists in the remote repository.
  * Branch name validation added to reject unsafe or malformed branch refs.

* **Tests**
  * Comprehensive unit tests covering branch resolution logic, remote validation behavior, and validation edge cases.

### CODE REVIEW

File: src/agentic_ci/branch.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+# Git ref validation regex from the git module
+_SAFE_REF_RE = re.compile(r"^[A-Za-z0-9._/\-~^]+$")
+
+
+class BranchResolutionError(Exception):
+    """Raised when branch resolution or validation fails critically."""
+
+
+def _validate_ref(name: str) -> bool:
```
Comment (mprpic) [inline src/agentic_ci/branch.py]: Why duplicate it here and not just import it?

File: src/agentic_ci/branch.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    Copied from git.py to maintain consistency with existing validation.
+    """
+    if not name or name.startswith("-"):
+        return False
+    if ".." in name or "@{" in name:
+        return False
+    return bool(_SAFE_REF_RE.match(name))
+
+
+def validate_branch_exists(repo_url: str, branch: str) -> bool:
```
Comment (mprpic) [inline src/agentic_ci/branch.py]: This probably belongs into the forge module.

File: src/agentic_ci/branch.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    except subprocess.TimeoutExpired:
+        log.warning("Branch validation timed out for %s branch %s", repo_url, branch)
+        return False
+    except (subprocess.CalledProcessError, FileNotFoundError) as exc:
+        log.debug("Branch validation failed for %s branch %s: %s", repo_url, branch, exc)
+        return False
+
+
+def resolve_branch(
```
Comment (mprpic) [inline src/agentic_ci/branch.py]: Let's use something more descriptive, like `resolve_branch_from_jira`.

File: src/agentic_ci/branch.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    ticket: dict[str, Any],
+    project_config: dict[str, Any],
+    component_config: dict[str, Any] | None = None,
+    repo_url: str | None = None,
+) -> str | None:
+    """Resolve target branch for a ticket using the resolution chain.
```
Comment (mprpic) [inline src/agentic_ci/branch.py]: Maybe:

Determine the target git branch to use by looking up version-specific information from a related Jira issue. This information is then mapped to project or component-specific configuration.

File: src/agentic_ci/branch.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        ticket.get("fixVersions", []),
+        component_config.get("name") if component_config else None,
+        repo_url,
+    )
+
+    # Step 1: Extract fixVersion[0]
+    fix_versions = ticket.get("fixVersions", [])
+    fix_version_name = None
+
+    if fix_versions and isinstance(fix_versions, list) and len(fix_versions) > 0:
```
Comment (mprpic) [inline src/agentic_ci/branch.py]: Could this ever not be a list?

File: src/agentic_ci/branch.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        return False
+    except (subprocess.CalledProcessError, FileNotFoundError) as exc:
+        log.debug("Branch validation failed for %s branch %s: %s", repo_url, branch, exc)
+        return False
+
+
+def resolve_branch(
+    ticket: dict[str, Any],
+    project_config: dict[str, Any],
+    component_config: dict[str, Any] | None = None,
```
Comment (mprpic) [inline src/agentic_ci/branch.py:23]: This entire function assumes some project and component level configuration, but I don't think we have the requirements of what these configs are expected to contain anywhere. Maybe that part from the autofix repo should be moved here? How do we create a reusable set of configuration schemas for some of these things?

File: src/agentic_ci/branch.py (hunk 1 above)
Comment (EmilienM) [reply src/agentic_ci/branch.py]: Fixed. Importing `_validate_ref` from `git.py` now.

File: src/agentic_ci/branch.py (hunk 2 above)
Comment (EmilienM) [reply src/agentic_ci/branch.py]: `validate_branch_exists` uses `git ls-remote` (subprocess), not the forge REST APIs. It fits better in the git CLI layer alongside `clone_repo`, `push_branch`, etc. Happy to move it to `git.py` if you'd prefer it out of `branch.py`, but forge seems like the wrong home for a subprocess call.

File: src/agentic_ci/branch.py (hunk 3 above)
Comment (EmilienM) [reply src/agentic_ci/branch.py]: Done, renamed to `resolve_branch_from_jira`.

File: src/agentic_ci/branch.py (hunk 4 above)
Comment (EmilienM) [reply src/agentic_ci/branch.py]: Updated the docstring along those lines.

File: src/agentic_ci/branch.py (hunk 5 above)
Comment (EmilienM) [reply src/agentic_ci/branch.py]: No, the Jira API always returns it as a list. Removed the check.

File: src/agentic_ci/branch.py (hunk 6 above)
Comment (EmilienM) [reply src/agentic_ci/branch.py:23]: The config schema lives in autofix (project-specific), not agentic-ci (reusable library). The split is intentional: agentic-ci provides the resolution logic, autofix owns the schema and wiring. Happy to discuss further if you think there's a better boundary.

File: src/agentic_ci/branch.py (hunk 2 above)
Comment (mprpic) [reply src/agentic_ci/branch.py]: ah, right, git.py is probably better then.

Comment (EmilienM) [reply src/agentic_ci/branch.py]: Moved `validate_branch_exists` to `git.py`. `branch.py` now imports it from there.

File: src/agentic_ci/branch.py (hunk 6 above)
Comment (mprpic) [reply src/agentic_ci/branch.py:23]: Right, but `project_config` here is a dict that we expect to have some specific attributes, but nowhere in here do we document what those should be apart from the fixtures in tests. So while the schema lives in autofix, that is invisible to any other consumer.

Comment (EmilienM) [reply src/agentic_ci/branch.py:23]: Good point. Removed `project_config` entirely since it was unused. `resolve_branch_from_jira()` now takes just `ticket`, `component_config`, and `repo_url`.

**Review by mprpic — APPROVED** [review]
